### PR TITLE
RT5682 sdw unified mach

### PR DESCRIPTION
**Pull request by naveen-manohar**
Patch for Debug_Purposes..
Based on https://github.com/thesofproject/linux/pull/1889

### CODE REVIEW

**Comment by naveen-manohar** [comment]
Apologize for posting RT5682 machine driver patches before PR:1889 Merge.

Wanted to discuss and get guidance about the following error that I observe with the patch-rebased to PR:1889, the same issue was not observed with PR:1885
1. Jack Detection works fine
2. Observe following error during playback
LOG
[   94.622100] intel-sdw sdw-master-0: failed to set stream pointer on codec dai rt5682-aif1
[   94.622108] intel-sdw sdw-master-0: ASoC: can't open interface SDW0 Pin2: -524
[   94.622116]  SDW0-Playback: ASoC: BE open failed -524
[   94.622121]  SDW0-headphone: ASoC: failed to start some BEs -524
[   94.622128]  SDW0-headphone: ASoC: BE playback disconnect check for SDW0-Playback
[   94.622130]  SDW0-headphone: freed DSP playback path SDW0-headphone -> SDW0-Playback
[   94.622258] rt5682 sdw:0:25d:5682:0: Parity error detected
[   94.622519] intel-sdw sdw-master-0: Slave status change
[  112.360613]  SDW0-headphone: ASoC: found 0 audio playback paths
[  112.360619]  SDW0-headphone: ASoC: SDW0-headphone no valid playback route
[  112.360623] sof_sdw sof_sdw: ASoC: find BE for widget ALH2.OUT
[  112.360625] sof_sdw sof_sdw: ASoC: try BE : ALH2.OUT
[  112.360628]  SDW0-headphone: connected new DPCM playback path SDW0-headphone -> SDW0-Playback
[  112.360651]  SDW0-headphone: ASoC: found 1 new BE paths
[  112.360653]  SDW0-Playback: ASoC: open playback BE SDW0-Playback

File: sound/soc/intel/boards/sof_sdw_rt5682.c [hunk 1]
```diff
@@ -0,0 +1,136 @@
+// SPDX-License-Identifier: GPL-2.0
+// Copyright (c) 2020 Intel Corporation
+
+/*
+ *  sof_sdw_rt5682 - Helpers to handle RT5682 from generic machine driver
+ */
+
+#include "sof_sdw_common.h"
```
Comment (lyakh) [inline sound/soc/intel/boards/sof_sdw_rt5682.c:16]: I think in its latest version this header isn't including all the headers "commonly needed" by all, each .c file is now expected to include all the appropriate headers itself. See #1889

Comment (naveen-manohar) [reply sound/soc/intel/boards/sof_sdw_rt5682.c:16]: > I think in its latest version this header isn't including all the headers "commonly needed" by all, each .c file is now expected to include all the appropriate headers itself. See #1889

Sure Shall correct this..  Thanks for Pointing this out.

Comment (naveen-manohar) [reply sound/soc/intel/boards/sof_sdw_rt5682.c:16]: > Apologize for posting RT5682 machine driver patches before PR:1889 Merge.
> 
> Wanted to discuss and get guidance about the following error that I observe with the patch-rebased to PR:1889, the same issue was not observed with PR:1885
> 
> 1. Jack Detection works fine
> 2. Observe following error during playback
>    LOG
>    [   94.622100] intel-sdw sdw-master-0: failed to set stream pointer on codec dai rt5682-aif1
>    [   94.622108] intel-sdw sdw-master-0: ASoC: can't open interface SDW0 Pin2: -524
>    [   94.622116]  SDW0-Playback: ASoC: BE open failed -524
>    [   94.622121]  SDW0-headphone: ASoC: failed to start some BEs -524
>    [   94.622128]  SDW0-headphone: ASoC: BE playback disconnect check for SDW0-Playback
>    [   94.622130]  SDW0-headphone: freed DSP playback path SDW0-headphone -> SDW0-Playback
>    [   94.622258] rt5682 sdw:0:25d:5682:0: Parity error detected
>    [   94.622519] intel-sdw sdw-master-0: Slave status change
>    [  112.360613]  SDW0-headphone: ASoC: found 0 audio playback paths
>    [  112.360619]  SDW0-headphone: ASoC: SDW0-headphone no valid playback route
>    [  112.360623] sof_sdw sof_sdw: ASoC: find BE for widget ALH2.OUT
>    [  112.360625] sof_sdw sof_sdw: ASoC: try BE : ALH2.OUT
>    [  112.360628]  SDW0-headphone: connected new DPCM playback path SDW0-headphone -> SDW0-Playback
>    [  112.360651]  SDW0-headphone: ASoC: found 1 new BE paths
>    [  112.360653]  SDW0-Playback: ASoC: open playback BE SDW0-Playback

Many Thanks @RanderWang for pointing out the below miss.
Code:
```
static struct sof_sdw_codec_info codec_info_list[] = {
        {
                                .id = 0x5682,
                .direction = {true, true},
-               .dai_name = "rt5682-aif1",
+               .dai_name = "rt5682-sdw",

```
Now I did a Loopback and was able to capture the audio. Shall update the PR and wait for PR:1889 merge.

File: sound/soc/intel/boards/sof_sdw_rt5682.c [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+
+	sdw_dev = bus_find_device_by_name(&sdw_bus_type, NULL, sdw_dev_name);
+	if (!sdw_dev)
+		return -EPROBE_DEFER;
+
+	ret = device_add_properties(sdw_dev, props);
+	put_device(sdw_dev);
+
+	return ret;
+}
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_sdw_rt5682.c]: that doesn't seem useful at all?

File: sound/soc/intel/common/soc-acpi-intel-tgl-match.c [hunk 3]
```diff
@@ -79,6 +87,15 @@ static const struct snd_soc_acpi_link_adr tgl_rvp[] = {
 	{}
 };
 
+static const struct snd_soc_acpi_link_adr tgl_ref[] = {
```
Comment (plbossart) [inline sound/soc/intel/common/soc-acpi-intel-tgl-match.c]: tgl_chromebook_base?

File: sound/soc/intel/boards/sof_sdw_rt5682.c (hunk 2 above)
Comment (naveen-manohar) [reply sound/soc/intel/boards/sof_sdw_rt5682.c]: Understood.. shall correct.

File: sound/soc/intel/common/soc-acpi-intel-tgl-match.c (hunk 3 above)
Comment (naveen-manohar) [reply sound/soc/intel/common/soc-acpi-intel-tgl-match.c]: Done.. Thanks for the review.

> tgl_chromebook_base?

Done.. Thanks for the review. re-posted.

**Comment by plbossart** [comment]
@naveen-manohar please rebase on top of topic/sof-dev, thanks

File: sound/soc/intel/boards/Kconfig [hunk 4]
```diff
@@ -554,6 +554,8 @@ config SND_SOC_INTEL_SOUNDWIRE_SOF_MACH
 	select SND_SOC_RT1308_SDW
 	select SND_SOC_RT1308
 	select SND_SOC_RT715_SDW
+	select SND_SOC_RT5682_SDW
+	select SND_SOC_RT5682
```
Comment (plbossart) [inline sound/soc/intel/boards/Kconfig]: the last line is unnecessary

````
config SND_SOC_RT5682_SDW
	tristate "Realtek RT5682 Codec - SDW"
	depends on SOUNDWIRE
	select SND_SOC_RT5682
	select REGMAP_SOUNDWIRE
````

Comment (naveen-manohar) [reply sound/soc/intel/boards/Kconfig]: Agreed on Kconfig..
rebased to sof-dev and re-verified functionality..

> minor change in the Kconfig, otherwise ok once it's rebased
Agreed on Kconfig..
rebased to sof-dev and re-verified functionality..

**Comment by naveen-manohar** [comment]
> @naveen-manohar please rebase on top of topic/sof-dev, thanks

Done.. Thanks

File: sound/soc/intel/boards/sof_sdw_rt5682.c [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+	jack = &ctx->sdw_headset;
+
+	snd_jack_set_key(jack->jack, SND_JACK_BTN_0, KEY_PLAYPAUSE);
+	snd_jack_set_key(jack->jack, SND_JACK_BTN_1, KEY_VOICECOMMAND);
+	snd_jack_set_key(jack->jack, SND_JACK_BTN_2, KEY_VOLUMEUP);
+	snd_jack_set_key(jack->jack, SND_JACK_BTN_3, KEY_VOLUMEDOWN);
+
+	ret = snd_soc_component_set_jack(component, jack, NULL);
+
```
Comment (lyakh) [inline sound/soc/intel/boards/sof_sdw_rt5682.c:103]: *if* something forces you to make another revision, would be good to remove this empty line: it's common to put lines like:
```
ret = fn(x);
if (ret < 0)
	return ret;
```
close together.

File: sound/soc/intel/boards/sof_sdw_rt5682.c [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+	ret = snd_soc_dapm_new_controls(&card->dapm, rt5682_widgets,
+					ARRAY_SIZE(rt5682_widgets));
+	if (ret) {
+		dev_err(card->dev, "rt5682 widgets addition failed: %d\n", ret);
+		return ret;
+	}
+
+	ret = snd_soc_dapm_add_routes(&card->dapm, rt5682_map,
+				      ARRAY_SIZE(rt5682_map));
+
```
Comment (lyakh) [inline sound/soc/intel/boards/sof_sdw_rt5682.c:76]: see comment below

File: sound/soc/intel/boards/sof_sdw_rt5682.c (hunk 5 above)
Comment (kv2019i) [reply sound/soc/intel/boards/sof_sdw_rt5682.c:103]: Thanks @lyakh I think this is a good reminder. If indeed a new revision is needed (I had one other comment), then better to change. Otherwise, I'm ok to merge this version.

File: sound/soc/intel/boards/sof_sdw.c [hunk 7]
```diff
@@ -72,6 +72,14 @@ static const struct dmi_system_id sof_sdw_quirk_table[] = {
...
 		.driver_data = (void *)SOF_SDW_PCH_DMIC,
 	},
+	{
+		.callback = sof_sdw_quirk_cb,
+		.matches = {
+			DMI_MATCH(DMI_SYS_VENDOR, "Google"),
+			DMI_MATCH(DMI_PRODUCT_NAME, "Volteer"),
+		},
+		.driver_data = (void *)(SOF_SDW_TGL_HDMI | SOF_SDW_PCH_DMIC),
+	},
```
Comment (kv2019i) [inline sound/soc/intel/boards/sof_sdw.c:82]: @naveen-manohar Hmm, patch looks good to go, but this change does not really match with the commit message. Shouldn't this be a separate commit? @plbossart  ?

File: sound/soc/intel/boards/sof_sdw_rt5682.c (hunk 5 above)
Comment (naveen-manohar) [reply sound/soc/intel/boards/sof_sdw_rt5682.c:103]: > _if_ something forces you to make another revision, would be good to remove this empty line: it's common to put lines like:
> 
> ```
> ret = fn(x);
> if (ret < 0)
> 	return ret;
> ```
> 
> close together.

Done.. Thank you

File: sound/soc/intel/boards/sof_sdw.c (hunk 7 above)
Comment (naveen-manohar) [reply sound/soc/intel/boards/sof_sdw.c:82]: > @naveen-manohar Hmm, patch looks good to go, but this change does not really match with the commit message. Shouldn't this be a separate commit? @plbossart ?

Thanks. Modified commit-message as suggested to suit the code.

File: sound/soc/intel/boards/sof_sdw_rt5682.c [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+			ret);
+		return ret;
+	}
+
+	jack = &ctx->sdw_headset;
+
+	snd_jack_set_key(jack->jack, SND_JACK_BTN_0, KEY_PLAYPAUSE);
+	snd_jack_set_key(jack->jack, SND_JACK_BTN_1, KEY_VOICECOMMAND);
+	snd_jack_set_key(jack->jack, SND_JACK_BTN_2, KEY_VOLUMEUP);
+	snd_jack_set_key(jack->jack, SND_JACK_BTN_3, KEY_VOLUMEDOWN);
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_sdw_rt5682.c:100]: add newline here, now it's harder to read the code
I think you removed the wrong newline

Comment (naveen-manohar) [reply sound/soc/intel/boards/sof_sdw_rt5682.c:100]: > add newline here, now it's harder to read the code
> I think you removed the wrong newline

Thanks @plbossart , aligned to same style as other codec helpers.

**Comment by kv2019i** [comment]
Ok, if the newlines are ok, let's go ahead and merged. Only known CI failures in the report.
